foreshadow/transformers/transformers.py

```diff
@@ -58,187 +92,288 @@ def wrap_transformer(transformer):
-                tuple(
-                    [
-                        s if type(s).__name__ != "type" else None
-                        for s in (argspec[3] if argspec[3] is not None else [])
-                    ]
-                ),
+    # use the same base
+    # metaclass as the transformer, otherwise we will get MRO metaclass
+    # issues in DFTransformer if we try to choose the base class for our
+    # metaclass that is not the same one for the transformer we are also
+    # extending.
+    class DFTransformerMeta(type(transformer)):
+        """Metaclass for DFTransformer to appear as parent Transformer."""
+
+        def __new__(mcs, *args, **kwargs):
+            class_ = super(DFTransformerMeta, mcs).__new__(
+                mcs, *args, **kwargs
             )
-            + argspec[4:]
-        )
-
-        # Copies keyword arguments and defaults
-        newargspec = (
-            (argspec[0] + tgt_argspec[0][1:],)
-            + argspec[1:4]
-            + (tgt_argspec[4], tgt_argspec[5])
-            + argspec[6:]
-        )
-
-        # Write new function
-        sigcall = inspect.formatargspec(
-            formatvalue=lambda val: "", *newargspec
-        )[1:-1]
-        signature = inspect.formatargspec(*newargspec)[1:-1]
-
-        signature = signature.replace("*,", "")
-        sigcall = sigcall.replace("*,", "")
-
-        new_func = (
-            "def _wrapper_(%(signature)s):\n"
-            "    return %(tgt_func)s(%(sigcall)s)"
-            % {
-                "signature": signature,
-                "tgt_func": "tgt_func",
-                "sigcall": sigcall,
-            }
-        )
-
-        # Set new metadata
-        evaldict = {"tgt_func": tgt_func}
-        exec(new_func, evaldict)
-        wrapped = evaldict["_wrapper_"]
-        wrapped.__name__ = name
-        wrapped.__doc__ = self.src_doc
-        wrapped.__module__ = tgt_func.__module__
-        wrapped.__dict__ = tgt_func.__dict__
-        return wrapped
-
-
-def init_partial(func):  # noqa: D202
-    """Partial function for injecting custom args into transformers.
-
-    Args:
-        func: function to wrap
-
-    Returns:
-        wrapped function
-
-    """
-
-    def transform_constructor(
-        self, *args, keep_columns=False, name=None, **kwargs
-    ):
-
-        self.name = args[-1]
-        self.keep_columns = args[-2]
-        func(self, *args[:-2], **kwargs)
-
-    return transform_constructor
-
-
-def pandas_partial(func):  # noqa: D202
-    """Partial function for the pandas transformer wrapper.
-
-    Args:
-        func: function to wrap
-
-    Returns:
-        wrapped function
-
-    """
-
-    @wraps(func)
-    def wrapper(self, *args, **kwargs):
-        return pandas_wrapper(self, func, *args, **kwargs)
-
-    return wrapper
+            class_.__name__ = transformer.__name__
+            class_.__doc__ = transformer.__doc__
+            return class_
+
+    class DFTransformer(transformer, metaclass=DFTransformerMeta):
```

Comment:
As with the DFTransformer class, does it need to be defined inside the wrapper method? I vaguely remember the reason but wanted to double check.

Reply:
That would also expose this to the documentation if need be.

Reply:
Yeah, unfortunately needs to extend from the transformer and that metaclass

Reply:
Yes, see previous.